eng/scripts/live-test-resource-cleanup.ps1

```diff
@@ -38,58 +48,167 @@ param (
+      return $true
+  }
+  return $false
+}
 
-foreach ($rg in $toDelete)
-{
-  if ($Force -or $PSCmdlet.ShouldProcess("$($rg.ResourceGroupName) (UTC: $($rg.Tags.DeleteAfter))", "Delete Group")) {
-    # Add purgeable resources that will be deleted with the resource group to the collection.
-    $purgeableResourcesFromRG = Get-PurgeableGroupResources $rg.ResourceGroupName
+function HasExpiredDeleteAfterTag([object]$ResourceGroup) {
+  $deleteDate = $ResourceGroup.Tags.DeleteAfter -as [DateTime]  # null if not exists
+  return $deleteDate -and [datetime]::UtcNow -gt $deleteDate
+}
 
-    if ($purgeableResourcesFromRG) {
-      $purgeableResources += $purgeableResourcesFromRG
-      Write-Verbose "Found $($purgeableResourcesFromRG.Count) potentially purgeable resources in resource group $($rg.ResourceGroupName)"
+function FindOrCreateDeleteAfterTag {
+  [CmdletBinding(SupportsShouldProcess = $true, ConfirmImpact = 'Medium')]
+  param(
+    [object]$ResourceGroup
+  )
+
+  if (!$ResourceGroup.Tags.DeleteAfter -as [datetime]) {
+    $deleteAfter = [datetime]::UtcNow.AddHours($DeleteAfterHours)
+    Write-Host "Adding DeleteAfer tag with value '$deleteAfter' to group '$(ResourceGroup.ResourceGroupName)'"
+    if ($Force -or $PSCmdlet.ShouldProcess("$($rg.ResourceGroupName) (UTC: $($rg.Tags.DeleteAfter))", "Update Group")) {
+      $ResourceGroup | Update-AzTag -Operation Merge -Tag @{ DeleteAfter = $deleteAfter }
     }
-    Write-Verbose "Deleting group: $($rg.ResourceGroupName)"
-    Write-Verbose "  tags $($rg.Tags | ConvertTo-Json -Compress)"
-    Write-Host ($rg | Remove-AzResourceGroup -Force -AsJob).Name
   }
 }
 
-# Purge all the purgeable resources and get a list of resources (as a collection) we need to follow-up on.
-Write-Host "Attempting to purge $($purgeableResources.Count) resources."
-$failedResources = @(Remove-PurgeableResources $purgeableResources -PassThru)
-if ($failedResources) {
-  Write-Warning "Timed out deleting the following $($failedResources.Count) resources. Please file an IcM ticket per resource type."
-  $failedResources | Sort-Object AzsdkResourceType, AzsdkName | Format-Table -Property @{l='Type'; e={$_.AzsdkResourceType}}, @{l='Name'; e={$_.AzsdkName}}
+function HasDeleteLock() {
+  return $false
+}
+
+function DeleteOrUpdateResourceGroups() {
+  [CmdletBinding(SupportsShouldProcess = $true, ConfirmImpact = 'Medium')]
+  param()
+
+  Write-Verbose "Fetching groups"
+  $allGroups = @(Get-AzResourceGroup)
+  $toDelete = @()
+  $toUpdate = @()
+  Write-Host "Total Resource Groups: $($allGroups.Count)"
+
+  foreach ($rg in $allGroups) {
+    if (HasExpiredDeleteAfterTag $rg) {
+      $toDelete += $rg
+    } elseif (HasValidOwnerTag $rg -or HasDeleteLock $rg) {
+      continue
+    } else {
+      $toUpdate += $rg
+    }
+  }
+
+  foreach ($rg in $toUpdate) {
+    FindOrCreateDeleteAfterTag $rg
+  }
+
+  # Get purgeable resources already in a deleted state.
+  $purgeableResources = @(Get-PurgeableResources)
```

Comment:
Do we need to fetch purgable resources if they aren't later deleted?

Reply:
I uncommented the purge line, so they should be purged now.